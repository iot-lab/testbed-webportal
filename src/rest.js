import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
