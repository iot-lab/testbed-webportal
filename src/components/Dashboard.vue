<template>
<div>
<div class="alert alert-info rounded-0" v-if="auth.isAdmin && nbPendingUsers > 0" style="font-weight: 300">
  <router-link :to="{name:'users'}" tag="div" class="container cursor">
    <i class="fa fa-exclamation-triangle" aria-hidden="true"></i>
    {{nbPendingUsers}} pending user accounts awaiting validation
  </router-link>
</div>
<div class="container mt-3">
  <div class="row">
    <div class="col-md-9">
      <h2>My experiments</h2>
      <p v-if="total.running != undefined">
        <span class="badge badge-pill badge-success">{{total.running}}</span> Running
        <span class="badge badge-pill badge-warning">{{total.upcoming}}</span> Scheduled
        <span class="badge badge-pill badge-dark">{{total.terminated}}</span> Completed
      </p>
      <p v-else>
        <i class="fa fa-spinner fa-spin fa-fw"></i>
      </p>
      <template v-if="total.running + total.upcoming">
        <h4 class="text-secondary">Scheduled</h4>
        <experiment-list user="@self" state="all_scheduled" :total="total.running + total.upcoming"></experiment-list>        
      </template>
      <p>
        <router-link :to="{name:'experiment'}" class="btn btn-primary">New experiment</router-link>
      </p>
      <template v-if="total.terminated">
        <h4 class="text-secondary">Recent</h4>
        <experiment-list user="@self" state="all_terminated" :total="total.terminated"></experiment-list>
      </template>
    </div>
    <div class="col">
      <h2>Platform status</h2>
      <p v-if="sites">
        <span class="badge badge-pill" :class="{'badge-primary': currentSite === 'all', 'badge-secondary': currentSite !== 'all'}" @click="currentSite = 'all'" style="cursor: pointer">{{sites.length}} sites</span>
        <span v-for="site in sites" class="badge badge-pill" :class="{'badge-primary': currentSite === site, 'badge-secondary': currentSite !== site}" style="margin-right: 4px; cursor: pointer"
        @click="currentSite = site">{{site.site}}</span>
      </p>
      <!-- <p v-if="stats.nodes">
          <span class="badge badge-pill badge-success">{{stats.nodes.Alive}}</span> nodes available
          <span class="badge badge-pill badge-warning">{{stats.nodes.Busy}}</span> busy
          <span class="badge badge-pill badge-danger">{{stats.nodes.Unavailable}}</span> unavailable
      </p> -->
      <p v-if="resources">
        <span class="badge badge-pill badge-success">{{getNodesCount(currentSite, ['Alive'])}}</span> nodes available
        <span class="badge badge-pill badge-warning">{{getNodesCount(currentSite, ['Busy'])}}</span> busy
        <span class="badge badge-pill badge-danger">{{getNodesCount(currentSite, ['Absent','Suspected'])}}</span> unavailable
      </p>
      <p v-else>
        <i class="fa fa-spinner fa-spin fa-fw"></i>
      </p>
      <p>
        <a class="btn btn-light" href="" @click.prevent="this.alert('todo')">Check future availability</a>
      </p>
    </div>
  </div>

</div>

</div> <!-- container -->

</template>

<script>
import ExperimentList from '@/components/parts/ExperimentList'
import {iotlab} from '@/rest'
import {auth} from '@/auth'

export default {
  name: 'Dashboard',

  components: {ExperimentList},

  data () {
    return {
      total: {},
      // stats: {},
      resources: [],
      sites: [],
      experiments: [],
      currentSite: 'all',
      auth: auth,
      nbPendingUsers: 0,
    }
  },

  created () {
    iotlab.getUserExperimentsCount().then(data => { this.total = data })
    // iotlab.getStats().then(data => { this.stats = data })
    iotlab.getSites().then(data => { this.sites = data })
    iotlab.getSiteResources().then(data => { this.resources = data })
    if (auth.isAdmin) {
      iotlab.getUsers().then(data => { this.nbPendingUsers = data.length })
    }
  },

  methods: {
    getNodesCount (site, stateList) {
      let siteList = (site === 'all') ? this.sites.map(key => key.site) : [site.site]
      return this.resources.filter(node => siteList.includes(node.site)).filter(node => stateList.includes(node.state)).length
    },
  },

}
</script>
