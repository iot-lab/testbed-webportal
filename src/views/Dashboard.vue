<template>
<div class="container mt-3">
  <h4 class="d-none d-sm-block d-md-none">My experiments</h4>
  <p v-if="total.running != undefined">
    <span class="badge badge-pill badge-success">{{total.running}}</span> Running
    <span class="badge badge-pill badge-warning">{{total.upcoming}}</span> Scheduled
    <span class="badge badge-pill badge-dark">{{total.terminated}}</span> Completed
  </p>
  <p v-else>
    <span class="badge badge-pill badge-success">?</span> Running
    <span class="badge badge-pill badge-warning">?</span> Scheduled
    <span class="badge badge-pill badge-dark">?</span> Completed
  </p>
  <experiment-list ref="runningExpList" title="Scheduled" user="@self" state="all_scheduled" @completed="updateTotal"></experiment-list>
  <p>
    <router-link :to="{name:'experiment'}" class="btn btn-primary">New experiment</router-link>
  </p>
  <template v-if="total.terminated">
    <experiment-list title="Recent" user="@self" state="all_terminated" :show="5" :total="total.terminated" @started="refreshRunning" @loaded="spinner = false"></experiment-list>
  </template>
  <template v-if="spinner">
    <i class="fa fa-spinner fa-spin fa-fw mr-1"></i>
    <i>loading experiments</i>
  </template>
</div> <!-- container -->

</template>

<script>
import ExperimentList from '@/components/ExperimentList'
import { iotlab } from '@/rest'
// import { sleep } from '@/utils'

export default {
  name: 'Dashboard',

  components: {ExperimentList},

  data () {
    return {
      total: {},
      spinner: true,
      created: false,
    }
  },

  beforeRouteEnter (to, from, next) {
    // refresh data when reentering the dashboard
    next(vm => {
      if (!vm.created) return // do not refresh until initial render as be done in created()
      vm.$notify({group: 'popup', clean: true})
      vm.$refs.runningExpList.refresh()
      vm.$refs.runningExpList.createPolling()
      vm.updateTotal()
    })
  },

  beforeRouteLeave (to, from, next) {
    // Indicate to the SubComponent to stop experiments polling
    this.$refs.runningExpList.destroyPolling()
    // Make sure to always call the next function, otherwise the hook will never be resolved
    // Ref: https://router.vuejs.org/en/advanced/navigation-guards.html
    next()
  },

  async created () {
    await this.updateTotal()
    this.created = true
  },

  async mounted () {
    this.$notify({group: 'popup', clean: true})
    let user = await iotlab.getUserInfo()
    if (!user.category) {
      this.$notify({group: 'popup', type: 'fill-user-profile'})
    }
  },

  methods: {
    async updateTotal () {
      // await sleep(2000)
      this.total = await iotlab.getUserExperimentsCount()
      // hide spinner as we are not going to fetch terminated experiments
      if (this.total.terminated === 0) this.spinner = false
    },
    refreshRunning () {
      this.$refs.runningExpList.refresh()
    },
  },

}
</script>
