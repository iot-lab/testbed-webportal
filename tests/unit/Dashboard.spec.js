import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Notifications from 'vue-notification'
import Dashboard from '@/views/Dashboard'
import { iotlab } from '@/rest'

jest.mock('@/rest')
iotlab.getSites = jest.fn().mockResolvedValue([{"site":"strasbourg"},{"site":"paris"},{"site":"grenoble"},{"site":"lille"},{"site":"saclay"},{"site":"lyon"}])
iotlab.getNodes = jest.fn().mockResolvedValue([])
iotlab.getRunningExperiments = jest.fn().mockResolvedValue([{"submitted_duration":60,"submission_date":"2018-08-29T23:14:27Z","nodes":["a8-1.devgrenoble.iot-lab.info","a8-4.devgrenoble.iot-lab.info","a8-17.devgrenoble.iot-lab.info","a8-3.devgrenoble.iot-lab.info","a8-5.devgrenoble.iot-lab.info","a8-7.devgrenoble.iot-lab.info","a8-9.devgrenoble.iot-lab.info","a8-6.devgrenoble.iot-lab.info","a8-2.devgrenoble.iot-lab.info","a8-8.devgrenoble.iot-lab.info"],"nb_nodes":10,"effective_duration":0,"name":"","stop_date":"1970-01-01T00:00:00Z","id":13226,"state":"Launching","scheduled_date":"2018-08-29T23:14:29Z","user":"bformet","start_date":"2018-08-29T23:14:29Z"}])
iotlab.getUserExperimentsCount = jest.fn().mockResolvedValue({"running":0,"terminated":9,"upcoming":0})

const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.use(Notifications)
const router = new VueRouter()

describe('Dashboard.vue', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(Dashboard, {localVue, router})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
