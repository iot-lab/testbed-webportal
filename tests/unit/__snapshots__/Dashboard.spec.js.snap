// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.vue should match snapshot 1`] = `
<div
  class="container mt-3"
>
  <h4
    class="d-none d-sm-block d-md-none"
  >
    My experiments
  </h4>
   
  <p>
    <span
      class="badge badge-pill badge-success"
    >
      ?
    </span>
     Running
    
    <span
      class="badge badge-pill badge-warning"
    >
      ?
    </span>
     Scheduled
    
    <span
      class="badge badge-pill badge-dark"
    >
      ?
    </span>
     Completed
  
  </p>
   
  <experimentlist-stub
    show="500"
    state="all_scheduled"
    step="40"
    title="Scheduled"
    total="0"
    user="@self"
  />
   
  <p>
    <router-link-stub
      class="btn btn-primary"
      event="click"
      tag="a"
      to="[object Object]"
    >
      New experiment
    </router-link-stub>
  </p>
   
  <!---->
   
  <i
    class="fa fa-spinner fa-spin fa-fw mr-1"
  />
   
  <i>
    loading experiments
  </i>
</div>
`;
